python_modules/dagster/dagster/_config/structured_config.py

```diff
@@ -132,7 +156,55 @@ def create_object_to_pass_to_user_code(self, context) -> Any:  # pylint: disable
         Default behavior for new class-based resources is to return itself, passing
         the actual resource object to user code.
         """
-        return self
+        return cast(T, self)
+
+    @classmethod
+    def partial(cls, **kwargs) -> "PartialResource[Self]":
+        """
+        Returns a partially initialized copy of the resource, with remaining config fields
+        set at runtime.
+        """
+        return PartialResource(cls, data=kwargs)
+
+    # Python descriptor
+    # https://docs.python.org/3/howto/descriptor.html
+    # Used to adjust the types of resource inputs and outputs, e.g. resource dependencies can be passed in
+    # as PartialResources or Resources, but will always be returned as Resources
+    def __get__(self, obj: "Resource", __owner: Any) -> Self:
+        ...
+
+    def __set__(self, obj: Optional[object], value: Union[T, "PartialResource[T]"]) -> None:
```

Comment:
I'd like some additional explanation here. I'm not following exactly how this works.

Reply:
A Python class implementing the [descriptor protocol](https://realpython.com/python-descriptors) (the `__get__` and `__set__` methods here) has custom behavior when attributes of that type are get or set. In this case, the `__get__` and `__set__` implementations determine what happens when a `Resource`-typed attribute of another class is get or set.

Our implementations themselves don't do anything interesting, but the type annotations on the methods determine the type hints affiliated with the data class constructor + attribute accesses:

```python
class MyClass(Resource):
    inner: Resource

# Type determined by __set__ param type hint, in this case Union[T, PartialResource[T]]
foo = MyClass(inner=...)

# Type determined by __get__ return value type hint, in this case T
foo.inner
```

(This is the protocol used, for example, to implement @property - our use-case is a lot simpler since we only use it to suggest to mypy/Pyright the right type to use)

Reply:
So this has _zero) impact on runtime behavior at all and this is only for type-checking purposes? These are not invoked at all at runtime?

Reply:
No, not that I can see for this use-case - w/ print methods inserted in the various descriptor methods, none trigger:

```python
>>> from dagster._config.structured_config import Resource
>>> class FooResource(Resource):
...     a_str: str
...
>>> class MyClass():
...     res: Resource
...
>>> x = MyClass()
>>> x.res = FooResource(a_str="foo")
>>> print(x.res.a_str)
foo
>>> class MyOuterResource(Resource):
...     inner: Resource
...
>>> outer = MyOuterResource(inner=FooResource(a_str="foo"))
>>> print(outer.inner.a_str)
foo
```

They can be manually invoked though:
```python
>>> x.res.__set_name__(x, "res")
__set_name__ invoked
>>> x.res.__get__(x, x).a_str
__get__ invoked
'foo'
```